### PR TITLE
Down sample perf

### PR DESCRIPTION
**Pull request by Yashg19**
- Created a separate program for down_sampling
- Updated __init__.py to add down sampling function
- Optimized down sampling code by changing _inv_index and _probe_index code
- Updated down sampling code to reflect Magellan manual with respect to parameters
- Added unit test cases for down sampling
- Added benchmark test cases for down sampling

@sanjibkd, @paulgc   - please do my code review

### CODE REVIEW

File: magellan/sampler/down_sample.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+
+    return h_table
+
+# down sample of two tables : based on sanjib's index based solution
+def down_sample(s_table, b_table, size, y):
+    """
+    This is down sample table function that down samples 2 tables A and B.
+
+    Args:
+        s_table: dataframe <input table A>
```
Comment (paulgc) [inline magellan/sampler/down_sample.py]: why not have tableA and tableB as input variables?
Also for docstring arguments the format is,
variable_name (type): description
Eg., tableA (dataframe): input table A

**Comment by sanjibkd** [comment]
@Yashg19 You might want to add try-except block in your benchmark test cases for the example data sets because none of them will be available with the package by default

**Comment by sanjibkd** [comment]
@Yashg19 Also try to pylint your code and try to achieve a score of at least 8

File: magellan/tests/test_sampler_down_sample.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+from magellan.io.parsers import read_csv_metadata
+
+datasets_path = os.sep.join([get_install_path(), 'datasets', 'test_datasets'])
+path_a = os.sep.join([datasets_path, 'restA.csv'])
+path_b = os.sep.join([datasets_path, 'restB.csv'])
+
+class DownSampleTestCases(unittest.TestCase):
+    def test_down_sample_table_valid_1(self):
+        A = read_csv_metadata(path_a)
+        B = read_csv_metadata(path_b, key='ID')
```
Comment (sanjibkd) [inline magellan/tests/test_sampler_down_sample.py]: Why not add a setUp() function for the class and then read the two tables there? See test_attr_equiv_blocker.py for reference.

File: magellan/sampler/down_sample.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+                "twelve", "twenty", "two", "un", "under", "until", "up", "upon", "us",
+                "very", "via", "was", "we", "well", "were", "what", "whatever", "when",
+                "whence", "whenever", "where", "whereafter", "whereas", "whereby",
+                "wherein", "whereupon", "wherever", "whether", "which", "while",
+                "whither", "who", "whoever", "whole", "whom", "whose", "why", "will",
+                "with", "within", "without", "would", "yet", "you", "your", "yours",
+                "yourself", "yourselves"
+    ]
+    return stop_words
+
```
Comment (paulgc) [inline magellan/sampler/down_sample.py]: @Yashg19  why not put stop words list in utils ? So that other modules can use it. I think overlap blocker may also use stop words.

Comment (Yashg19) [inline magellan/sampler/down_sample.py]: @paulgc  taken care of the same

File: magellan/tests/test_sampler_down_sample.py (hunk 2 above)
Comment (Yashg19) [inline magellan/tests/test_sampler_down_sample.py]: Taken care

File: magellan/sampler/down_sample.py (hunk 1 above)
Comment (Yashg19) [inline magellan/sampler/down_sample.py]: @paulgc  - taken care of your points

File: magellan/sampler/down_sample.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+import math
+import logging
+import random
+import numpy as np
+import pandas as pd
+import pyprind
+from magellan.utils.generic_helper import get_install_path
+import magellan.catalog.catalog_manager as cm
+
+LOGGER = logging.getLogger(__name__)
```
Comment (kvpradap) [inline magellan/sampler/down_sample.py]: It will be good to just use logger than LOGGER

File: magellan/sampler/down_sample.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+INSTALL_PATH = get_install_path()
```
Comment (kvpradap) [inline magellan/sampler/down_sample.py]: We will avoid global varaibles in general. You can call this function (get_install_path) where you use this global variable.

File: magellan/sampler/down_sample.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+DATASET_PATH = os.sep.join([INSTALL_PATH, 'utils'])
```
Comment (kvpradap) [inline magellan/sampler/down_sample.py]: Same comment as above

**Comment by Yashg19** [comment]
Added comments and changes as recommended by Pradap.

**Comment by Yashg19** [comment]
@sanjibkd  @paulgc  - please review if I have implemented all review comments given by you all :)

File: benchmarks/benchmark_down_sample_sampler.py [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+
+class TimeDownSampleRestaurants:
+    def time_down_sample_tables(self):
+        path_for_a = os.sep.join([DATASET_PATH, 'restaurants', 'A.csv'])
+        path_for_b = os.sep.join([DATASET_PATH, 'restaurants', 'B.csv'])
+        self.size = 500
+        self.y_param = 2
+        try:
+            self.A = read_csv_metadata(path_for_a)
+            self.B = read_csv_metadata(path_for_b)
```
Comment (sanjibkd) [inline benchmarks/benchmark_down_sample_sampler.py]: Time taken to read the tables will also get included in the down sampling time. You should separate the data loading time.

File: magellan/sampler/down_sample.py [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+    """
+    This is down sample table function that down samples 2 tables A and B. First it randomly selects size tuples
+    from the table B to be table B'. First step is to build an inverted index (token, tuple_id) on table A - say I.
+    For each tuple x ∈ B', the algorithm finds a set P of k/2 tuples from A (inverted index) that match x,
+    and a set Q of k/2 tuples randomly selected from A \ P. The idea is for A' and B' to share some matches yet be
+    as representative of A and B as possible.
+
+    Args:
+        table_a (dataframe): input table A
+        table_b (dataframe): input table B
```
Comment (sanjibkd) [inline magellan/sampler/down_sample.py]: Use (DataFrame). Refer to Pradap's docstrings for the convention.

File: magellan/sampler/down_sample.py [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+        str_val = str_val.difference(stop_words)
+
+        # building inverted index I from set of tokens
+        for token in str_val:
+            lst = inv_index.get(token, None)
+            if lst is None:
+                inv_index[token] = [pos]
+            else:
+                lst.append(pos)
+                inv_index[token] = lst
```
Comment (paulgc) [inline magellan/sampler/down_sample.py]: why do you need to assign the list back ?

File: magellan/sampler/down_sample.py [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+    pos = 0
+
+    # For each row in the dataframe of the input table, we will fetch all string values from string column indices
+    # and will concatenate them. Next step would be to tokenize them using set and remove all the stop words
+    # from the list of tokens. Once we have the list of tokens, we will iterate through the list of tokens
+    # to identify its position and will create an inverted index.
+    for row in table.itertuples():
+        str_val = ''
+        for list_item in str_cols_ix:
+            str_val += str(row[list_item + 1]).lower() + ' '
```
Comment (paulgc) [inline magellan/sampler/down_sample.py]: if you iterate using "itertuples(index=False)", you do not need to increment list_item

**Comment by Yashg19** [comment]
@paulgc  @sanjibkd  - all review comments are incorporated. I have tested the code but wasn't able to run nose test again to re-verify the changes. I have tested the code separately with different input parameters. 
While running nose test or asv run - I was getting the error:

import magellan as mg
--> from magellan.blocker.attr_equiv_blocker import AttrEquivalenceBlocker
-->-->from joblib import Parallel, delayed

ImportError: No module named joblib

@kvpradap  - copying you as the changes are ready to pull if there are no more review comments.
